tpu-client-next/src/node_address_service/leader_tpu_cache_service.rs

```diff
@@ -0,0 +1,517 @@
+pub struct LeaderTpuCacheService {
+    handle: Option<JoinHandle<Result<(), LeaderTpuCacheServiceError>>>,
+    cancel: CancellationToken,
+}
+
+/// Receiver for leader TPU socket address updates from
+/// [`LeaderTpuCacheService`].
+#[derive(Clone)]
+pub struct LeaderUpdateReceiver {
+    receiver: watch::Receiver<(Vec<SocketAddr>, bool)>,
+}
+
+impl LeaderUpdateReceiver {
+    pub fn next_leaders(&self, lookahead_leaders: usize) -> Vec<SocketAddr> {
+        let (leaders, extended) = self.receiver.borrow().clone();
+        let lookahead_leaders = if extended {
+            lookahead_leaders.saturating_add(1)
+        } else {
+            lookahead_leaders
+        };
+        extract_send_leaders(&leaders, lookahead_leaders)
+    }
+}
+
+impl LeaderTpuCacheService {
+    /// Run the [`LeaderTpuCacheService`], returning receiver and the service.
+    pub async fn run(
+        rpc_client: Arc<impl ClusterInfoProvider + 'static>,
+        mut slot_receiver: SlotReceiver,
+        config: LeaderTpuCacheServiceConfig,
+        cancel: CancellationToken,
+    ) -> Result<(LeaderUpdateReceiver, Self), LeaderTpuCacheServiceError> {
+        let (mut leader_tpu_map, mut epoch_info, mut slot_leaders) = Self::initialize_state(
+            rpc_client.as_ref(),
+            slot_receiver.clone(),
+            config.max_consecutive_failures,
+        )
+        .await?;
+        let (current_slot, lookahead_leaders) = get_slot_and_lookahead(
+            slot_receiver.slot(),
+            &slot_leaders,
+            config.lookahead_leaders,
+        );
+        let leaders = Self::leader_sockets(
+            current_slot,
+            lookahead_leaders,
+            &slot_leaders,
+            &leader_tpu_map,
+        );
+
+        let (leaders_sender, leaders_receiver) =
+            watch::channel((leaders, config.lookahead_leaders != lookahead_leaders));
+
+        let cancel_clone = cancel.clone();
+        let main_loop = async move {
+            let mut num_consequent_failures: usize = 0;
+            let mut refresh_tpu_interval = interval(config.refresh_nodes_info_every);
+            loop {
+                tokio::select! {
+                    _ = refresh_tpu_interval.tick() => {
+                        try_update(
+                            "cluster TPU ports",
+                            &mut leader_tpu_map,
+                            || LeaderTpuMap::new(rpc_client.as_ref()),
+                            &mut num_consequent_failures,
+                            config.max_consecutive_failures,
+                        ).await?;
+                        debug!("Updated cluster TPU ports");
+                    }
+                    res = slot_receiver.changed() => {
+                        debug!("Changed slot receiver");
+                        if let Err(e) = res {
+                            warn!("Slot receiver channel closed: {e}");
+                            break;
+                        }
```

Comment:
*minor*

Current `res` type is `Result<(), Error>`.
Would it make sense to make sure that if it changes, the compiler will flag it here?
In which case, you want a full `match`:

```suggestion
                        match res {
                            Ok(()) => (),
                            Err(e) => {
                                warn!("Slot receiver channel closed: {e}");
                                break;
                            }
                        }
```